libdevice/device_imf.hpp

```diff
@@ -0,0 +1,527 @@
+// pre-ci tests, the pre-ci environment used some legacy c++ std library
+// which doesn't include this function. Need to report to pre-ci owners.
+template <typename To, typename From>
+static inline constexpr To __bit_cast(const From &from) {
+  static_assert(sizeof(To) == sizeof(From),
+                "Can't do bit cast between 2 types with different sizes!");
+  /*static_assert(std::is_trivially_copyable<From>::value &&
+                    std::is_trivially_copyable<To>::value,
+                "Can't do bit cast for type which is not trivially
+     copyable!");*/
+  return __builtin_bit_cast(To, from);
+}
+
+#if defined(__LIBDEVICE_HOST_IMPL__)
+#include <cfenv>
+#pragma STDC FENV_ACCESS ON
+
+template <typename Tp> static inline Tp __double2Tp_host(double x, int rdMode) {
+  static_assert(std::is_same<Tp, float>::value ||
+                    std::is_same<Tp, int>::value ||
+                    std::is_same<Tp, unsigned int>::value ||
+                    std::is_same<Tp, long long int>::value ||
+                    std::is_same<Tp, unsigned long long int>::value,
+                "Invalid type for double conversion!");
+
+  const int roundingOriginal = fegetround();
+  fesetround(rdMode);
+  Tp res;
+  if (std::is_same<Tp, float>::value)
+    res = static_cast<float>(x);
+  else
+    res = static_cast<Tp>(__builtin_nearbyint(x));
+  fesetround(roundingOriginal);
+  return res;
+}
+
+template <typename Tp> static inline Tp __float2Tp_host(float x, int rdMode) {
+  static_assert(std::is_same<Tp, int>::value ||
+                    std::is_same<Tp, unsigned int>::value ||
+                    std::is_same<Tp, long long int>::value ||
+                    std::is_same<Tp, unsigned long long int>::value,
+                "Invalid type for float conversion!");
+
+  const int roundingOriginal = fegetround();
+  fesetround(rdMode);
+  Tp res = static_cast<Tp>(__builtin_nearbyintf(x));
+  fesetround(roundingOriginal);
+  return res;
+}
+
+template <typename TyINT, typename TyFP>
+static inline TyFP __integral2FP_host(TyINT x, int rdMode) {
+  static_assert((std::is_same<TyINT, int>::value ||
+                 std::is_same<TyINT, unsigned int>::value ||
+                 std::is_same<TyINT, long>::value ||
+                 std::is_same<TyINT, unsigned long>::value ||
+                 std::is_same<TyINT, long long int>::value ||
+                 std::is_same<TyINT, unsigned long long int>::value) &&
+                    (std::is_same<TyFP, float>::value ||
+                     std::is_same<TyFP, double>::value),
+                "Invalid integral to FP conversion!");
+  const int roundingOriginal = fegetround();
+  fesetround(rdMode);
+  TyFP res = static_cast<TyFP>(x);
+  fesetround(roundingOriginal);
+  return res;
+}
+#endif // __LIBDEVICE_HOST_IMPL__
+
+static inline float __fclamp(float x, float y, float z) {
+#if defined(__LIBDEVICE_HOST_IMPL__)
+  return __builtin_fmin(__builtin_fmax(x, y), z);
+#elif defined(__SPIR__)
```

Comment:
This and a bunch of the other definitions will fail if neither `__LIBDEVICE_HOST_IMPL__` nor `__SPIR__` is defined. Could this have a `static_assert` at the beginning of the file to warn in the case that this happens?

Reply:
Hi, @steffenlarsen 
Code in this file is guarded by another macro "__LIBDEVICE_IMF_ENABLED": https://github.com/intel/llvm/pull/5999/files#diff-9cbc9200719f1499b106121cbcb9010e4a8b5629983bd18e185696a8aea49d67R17
And this marco is defined in device.h when either __SPIR__ or __LIBDEVICE_HOST_IMPL__ is defined.
So, we can add a check in device_imf.hpp when the macro "__LIBDEVICE_IMF_ENABLED" is not defined, is that OK?

Thanks very much.

Reply:
My primary worry is if somehow we unintentionally break this guarantee in the future. We should hopefully get warnings, but I am hoping a `static_assert` at the top would help whoever breaks it understand the reason for the failures faster than having to dig through a pile of missing return errors.

Reply:
This has been addressed.